server/src/test/java/org/elasticsearch/monitor/jvm/HotThreadsTests.java

```diff
@@ -0,0 +1,291 @@
+            when(threadOne.getStackTrace()).thenReturn(testCase.one);
+
+            ThreadInfo threadTwo = mock(ThreadInfo.class);
+            when(threadTwo.getThreadName()).thenReturn("Thread Two");
+            when(threadTwo.getStackTrace()).thenReturn(testCase.two);
+
+            assertEquals(hotThreads.similarity(threadOne, threadTwo), testCase.similarityScore);
+        }
+
+        ThreadInfo threadOne = mock(ThreadInfo.class);
+        when(threadOne.getThreadName()).thenReturn("Thread One");
+        when(threadOne.getStackTrace()).thenReturn(testCases[0].one);
+
+        assertEquals(hotThreads.similarity(threadOne, null), 0);
+    }
+
+    // We call this helper for each different mode to reset the before and after timings.
+    private List<ThreadInfo> makeThreadInfoMocksHelper(ThreadMXBean mockedMXBean, long[] threadIds) {
+        List<ThreadInfo> allInfos = new ArrayList<>(threadIds.length);
+
+        for (var threadId : threadIds) {
+            // We first return 0 for all timings, then a true value to create the reporting deltas.
+            when(mockedMXBean.getThreadCpuTime(threadId)).thenReturn(0L).thenReturn(-threadId);
+            ThreadInfo mockedThreadInfo = mock(ThreadInfo.class);
+            when(mockedMXBean.getThreadInfo(eq(threadId), anyInt())).thenReturn(mockedThreadInfo);
+            when(mockedThreadInfo.getThreadName()).thenReturn(String.format(Locale.ROOT, "Thread %d", -threadId));
+
+            // We create some variability for the blocked and waited times. Odd and even.
+            when(mockedThreadInfo.getBlockedCount()).thenReturn(0L).thenReturn((-threadId) % 2);
+            long blockedTime = (((-threadId) % 2) == 0) ? 0L : -threadId;
+            when(mockedThreadInfo.getBlockedTime()).thenReturn(0L).thenReturn(blockedTime);
+
+            when(mockedThreadInfo.getWaitedCount()).thenReturn(0L).thenReturn((-threadId + 1) % 2);
+            long waitTime = (((-threadId + 1) % 2) == 0) ? 0L : -threadId;
+            when(mockedThreadInfo.getWaitedTime()).thenReturn(0L).thenReturn(waitTime);
+
+            when(mockedThreadInfo.getThreadId()).thenReturn(threadId);
+
+            var stack = makeThreadStackHelper(
+                List.of(
+                    new String[]{"org.elasticsearch.monitor.test", String.format(Locale.ROOT, "method_%d", (-threadId) % 2)},
+                    new String[]{"org.elasticsearch.monitor.testOther", "methodFinal"}
+                )).toArray(new StackTraceElement[0]);
+            when(mockedThreadInfo.getStackTrace()).thenReturn(stack);
+
+            allInfos.add(mockedThreadInfo);
+        }
+
+        return allInfos;
+    }
+
+    public void testInnerDetect() throws Exception {
+        ThreadMXBean mockedMXBean = mock(ThreadMXBean.class);
+        when(mockedMXBean.isThreadCpuTimeSupported()).thenReturn(true);
+
+        // Using negative thread ids to ensure they don't collide with our JUnit thread id.
```

Comment:
I don't understand this comment as we are mocking stuff now. Can you explain it?

Reply:
I see how my comment is unclear now :). I'll expand the comment with a better description. 

Essentially, the code inside innerDetect calls Thread.currentThread().getThreadId(), to exclude the HotThreads runner thread itself from the report. While it's highly unlikely that the Thread.currentThread().getThreadId() will return a number between [1, 4], it might happen and the test will become flaky, effectively it will fail to report one of the ids that matches the threadId of the current JUnit thread.

I could eliminate this issue by mocking Thread.currentThread(), however it's a static call and I'd need PowerMockito to mock it.

But... now that I think about it, the code only needs the current thread id and not really the current Thread reference. So with a small refactor to the caller arguments of innerDetect I can eliminate the static call, and supply non-matching thread id for the test. This will simplify all those -/+1 conversions too. This will be better, I'll go with the latter approach.

Thanks!

Reply:
Thank you for the explanation. I am fine with the current code.

Reply:
Your question was spot on, it got me thinking more about it and with the small change to pass the threadId I was able to add one more test. This new test ensures that the thread processing code does indeed skip the HotThreads request thread itself.